### PR TITLE
refactor: apply Clippy suggestions

### PR DESCRIPTION
**Pull request by aleksator**
This PR alongside with https://github.com/pubgrub-rs/pubgrub/pull/46 fixes all Clippy suggestions in the current state of the repo.
After both are merged, Clippy can be added to CI.

I've also taken a liberty to clean up small TODOs.

### CODE REVIEW

**Comment by Eh2406** [comment]
Lots of little things that each look good. This overall looks very good. I will try to revue one commit at time to ensure I like each thing.

**Comment by aleksator** [comment]
> Lots of little things that each look good. This overall looks very good. I will try to revue one commit at time to ensure I like each thing.

Yeah they are all atomic. Let me know if this is too much and I should pull some things out of this PR like I did with #46.

File: src/internal/partial_solution.rs
Comment (Eh2406) [inline src/internal/partial_solution.rs]: Why is this here instead of in `incompatibility.rs`?

Comment (aleksator) [reply src/internal/partial_solution.rs]: No reason besides the return type of functions I was trying to simplify was in `partial_solution.rs`. I will move it to `incompatibility.rs`, it does make more sense there since it's used in `Relation` enum.

Are you fine with it being a type alias rather than a separate type? I have a commit where I had it extracted into a struct, let me know if you would prefer that :smile:

**Review by Eh2406 — APPROVED** [review]
Just the one nit.

File: src/internal/partial_solution.rs
Comment (Eh2406) [reply src/internal/partial_solution.rs]: I have no methods to add to it so type alias works.

**Review by Dylan-DPC-zz — APPROVED** [review]

File: src/internal/partial_solution.rs
Comment (aleksator) [reply src/internal/partial_solution.rs]: Alright. Thank you spotting this! I changed the last commit.

**Comment by aleksator** [comment]
I've changed the last commit (moved `PackageTerm` to `incompatibility.rs`).

Would you prefer to have all of them squashed or kept separate? (squash merge vs rebase merge)

**Comment by Eh2406** [comment]
The commits are atomic so I'd be ok with rebase, but in general @mpizenberg prefers squash. So I'll leave it up to him.

File: src/internal/core.rs
Comment (mpizenberg) [inline src/internal/core.rs]: I'm not a big fan of the `pub(crate)` thing as you may have noticed. I generally think that if I need something to be pub, everybody needs it. And I think the round of `pub(crate)` we did on the `term` module because I didn't spend enought time on the reporting API will byte us for #40 .

What I meant here by "remove pub" was "the outside world (outside this module)" should not know how this is done internally. And in fact #21 solves already two of those pubs by re-organizing indices. So we should maybe leave those out of this PR

Comment (mpizenberg) [reply src/internal/core.rs]: In addition, the `pub(crate)` here is equivalent to the `pub` anyway since `internal::` is not `pub`.

File: src/range.rs
Comment (mpizenberg) [inline src/range.rs]: This feels a lot less readable to me. What is Clippy saying about this one?

File: src/internal/partial_solution.rs
Comment (mpizenberg) [inline src/internal/partial_solution.rs]: I'm not certain, but I believe you can write something like `|AssignmentDepth { assignment, .. }| `

Comment (mpizenberg) [reply src/internal/partial_solution.rs]: same for the change above on rposition

File: src/internal/partial_solution.rs
Comment (mpizenberg) [inline src/internal/partial_solution.rs]: Maybe `AssignmentWithDecisionLevel` would be more adequate even though a bit redundant with its content. `AssignmentDepth` sounds like something it's not. Is it the fact that `usize` does not convey a lot of meaning that bothers you? Maybe a type alias would be enough here?

File: src/internal/partial_solution.rs
Comment (mpizenberg) [inline src/internal/partial_solution.rs:22]: Alright just in the following commit :+1: 
I feel like a type alias would be enough though.

Comment (mpizenberg) [reply src/internal/partial_solution.rs:22]: But ok for the newtype

File: src/internal/partial_solution.rs
Comment (mpizenberg) [inline src/internal/partial_solution.rs]: These really should not be put together though in my opinion. They have no other reason to be together than being two outputs of a function.

Comment (mpizenberg) [reply src/internal/partial_solution.rs]: The logic inside the satisfier isn't something very stable either. so having the flexibility of easily changing the return types there is important.

File: src/internal/incompatibility.rs
Comment (mpizenberg) [inline src/internal/incompatibility.rs:49]: Ok

Comment (mpizenberg) [reply src/internal/incompatibility.rs:49]: But I don't see much value. `(P, Term<V>)` seems pretty straightforward

Comment (mpizenberg) [reply src/internal/incompatibility.rs:49]: Sometimes adding a level of indirection is not necessary

**Review by mpizenberg — COMMENTED** [review]
I have mixed feelings on a lot of things in this PR. I'm basically ok with all Clippy suggestions but not ok with some names used and some changes. Detailed in review items.

**Comment by Eh2406** [comment]
If there are a few lintes that are holding this PR back, then we should split them out and merge the clear cut parts.

File: src/range.rs
Comment (aleksator) [reply src/range.rs]:
```
warning: `if` chain can be rewritten with `match`
   --> src/range.rs:202:21
    |
202 | /                     if r2 < l1 {
203 | |                         right = right_iter.next();
204 | |                     } else if r2 == l1 {
205 | |                         segments.extend(right_iter.cloned());
...   |
211 | |                         break;
212 | |                     }
    | |_____________________^
    |
    = help: Consider rewriting the `if` chain to use `cmp` and `match`.
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#comparison_chain
```

Comment (aleksator) [reply src/range.rs]: I like it because it shows the intent: you are matching over all possible comparison values.

There's a principle I use while coding that can be formulated as "use the least powerful construct that does the job".

Like when you are iterating over values in a list, `foreach` is preferred over `for (int i = 0; i < len; ++i)` since the latter is more powerful construct, using that indicates that you probably want to manipulate indexes somehow. If you aren't doing that, `foreach` says "simple iteration".

Same deal here: match over possible comparison values says exactly that. If chain is more powerful, you would use it if you wanted to do something additional in the conditions and you have to check whether that is the case. It is not the case here, hence we can use less powerful tool and save ourselves condition checking.

File: src/internal/partial_solution.rs
Comment (aleksator) [reply src/internal/partial_solution.rs]: Or even better, we can just do `.map(|assignment_depth| assignment_depth.assignment)`. Thanks for bringing my attention to these lines.

File: src/internal/partial_solution.rs
Comment (aleksator) [reply src/internal/partial_solution.rs]: > Maybe AssignmentWithDecisionLevel would be more adequate even though a bit redundant with its content. 

I can rename that no problem. A bit verbose, but can't come up with anything better. `DecidedAssignment`? :smile: 

> Is it the fact that usize does not convey a lot of meaning that bothers you?

I like using type system to indicate meaning and make it easily greppable. Allows for easy type changes as well: say we want to reduce `u32` to `u16`, it would simplify doing such refactoring a lot.

> Maybe a type alias would be enough here?

Let's see, type alias also provides some of those benefits. There is a problem that you could forget to use it, so searching usages becomes less reliable. Even if you never forget new instance creation becomes hidden (just 0/1 in code). It also can not guard against adding that and something else.
As for benefits... Hm... Both are fully optimized away, so I don't know. Why do you like it more?

Comment (aleksator) [reply src/internal/partial_solution.rs]: > `DecidedAssignment`

`AppliedAssignment` as another option.

> Maybe a type alias would be enough here?

Wait, I see another thread talking about type alias/newtype for `DecisionLevel`. Were you asking about the whole `AssignmentDepth` here? If so, I got confused by "Is it the fact that usize does not convey a lot of meaning that bothers you?" question.

Let's see... It keeps return types in functions the same as newtype, quite readable. I'm only concerned about matching on it: you'd have to look up the type definition and do tuple destructuring in that case, while newtype allows field access by name.

File: src/internal/partial_solution.rs
Comment (aleksator) [reply src/internal/partial_solution.rs]: > The logic inside the satisfier isn't something very stable either. so having the flexibility of easily changing the return types there is important.

Doesn't having a grouping type make it more easily changeable? You could extend the struct/change inner fields without touching type signature.

> These really should not be put together though in my opinion. They have no other reason to be together than being two outputs of a function.

True, the purpose of the change was to provide some context on those return types.
Before:
```
 ) -> Option<(AssignmentDepth<P, V>, &'a [AssignmentDepth<P, V>])>
```
After:
```
 ) -> Option<AssignmentHistory<'a, P, V>>
```
I think looking at type signature of "before" version does not make it clear why those are separate and why AssignmentDepth is not inside the slice, and what that slice means.

Would you be okay with making this a type alias if you think a struct definition doesn't pull it's weight here?
I would like having one of those as a return type.

File: src/internal/incompatibility.rs
Comment (aleksator) [reply src/internal/incompatibility.rs:49]: I admit this was just to fix a clippy lint, both before and after are quite easy to read IMO. I weighted the options of adding a suppress for the lint, and defining a type alias for PackageTerm and went with the latter.

The reason was we already have this meaning implicitly defined in our library: we have `package_terms` map that we iterate over, and the iterator returns this tuple. So it didn't feel like a new thing you have to look up a definition of.

If you disagree with the reasoning, let me know and I will change it to a lint suppression instead.

**Comment by aleksator** [comment]
> If there are a few lintes that are holding this PR back, then we should split them out and merge the clear cut parts.

I think what was holding this PR back was me not working through the comments until today :)

We could merge a few uncontroversial changes. There are a lot of discussion points though, and I don't know how to move them to another PR. Gitlab has "resolve in another issue" button, but I don't see anything similar here.
Does anyone know?

Alternatively, we could fix everything in this PR since I have time this weekend to correct everything.

File: src/internal/core.rs
Comment (aleksator) [reply src/internal/core.rs]: Will revert.

File: src/internal/partial_solution.rs
Comment (aleksator) [reply src/internal/partial_solution.rs]: Changed both in the latest commit "refactor: shorten AssignmentDepth matching".

File: src/internal/partial_solution.rs
Comment (aleksator) [reply src/internal/partial_solution.rs:22]: > Alright just in the following commit +1

Sorry, I don't follow what you mean here :)

**Comment by aleksator** [comment]
> We could merge a few uncontroversial changes. There are a lot of discussion points though, and I don't know how to move them to another PR. Gitlab has "resolve in another issue" button, but I don't see anything similar here.
Does anyone know?

Just realized I could do the opposite: extract approved parts into a separate PR and leave all the questionable items here. Let me do just that.

File: src/internal/partial_solution.rs
Comment (mpizenberg) [reply src/internal/partial_solution.rs:22]: Just because I was reviewing one commit after the other. In the previous commit, I suggested giving a type alias for decision level and here in the following commit you created a type for that. Newtype is good :+1:

File: src/internal/partial_solution.rs
Comment (mpizenberg) [reply src/internal/partial_solution.rs]: > Doesn't having a grouping type make it more easily changeable?

Yeah you're right on this one XD.

> I think looking at type signature of "before" version does not make it clear why those are separate and why AssignmentDepth is not inside the slice, and what that slice means.

I'll try to think a bit more about that return type and come back with some ideas (or not)

File: src/internal/incompatibility.rs
Comment (mpizenberg) [reply src/internal/incompatibility.rs:49]: Alright, I guess both are ok. Choose what you prefer :)

**Review by mpizenberg — COMMENTED** [review]
Ok so my last issues in this PR are with the `AssignmentDepth` and `AssignmentHistory` types. I'll try to spend some time of this tomorrow or Monday.

File: src/internal/partial_solution.rs
Comment (aleksator) [reply src/internal/partial_solution.rs:22]: Ah, and I misunderstood that you were arguing against adding a newtype or type alias for `decision_level` and wanted to keep a plain integer type. Got it now :+1:

File: src/internal/incompatibility.rs
Comment (aleksator) [reply src/internal/incompatibility.rs:49]: Thanks. Let's keep a type alias then :)

**Comment by aleksator** [comment]
Summary:

`AssignmentDepth` could be named better:
- `AssignmentWithDecisionLevel`
- `DecidedAssignment`
- `AppliedAssignment`

`AssignmentHistory` - alright name, but questionable grouping.

**Comment by mpizenberg** [comment]
Regarding `AssignmentDepth`. The main purpose of this struct is to give a name to the integer recording the decision level associated with an Assignement. For context, initially the decision level was a field inside `Assignment` that used to be a struct. Then I extracted it out of the `Assignment` since an assignment doesn't have to know when it was created. This enabled some simplifications in the code structure, and `Assignment` switched from a struct to an enum.

When the decision level was extracted, it was then stored side-by-side with the assignment in a pair `(usize, Assignment)` inside a vec called `history` being the ordered history of all assignments created for the partial solution.

In this PR, we agreed to replace the `usize` (that should have been a `u32`) by the newtype `DecisionLevel(u32)` which is more appropriate. As such, I think the pair `(DecisionLevel, Assignment)` is perfect for the generic type for `history`. So I'd go for `history: Vec<(DecisionLevel, Assignment<P, V>)>`. But if you really think a struct would be better, I'd suggest something like `AssignmentRecord`. History is after all a big record of "dated" events. Here the "date" is the decision level.

**Comment by mpizenberg** [comment]
The `find_satisfier` and `find_previous_satisfier` methods have had fluctuating interfaces. At the beginning they returned the same type, but after few refactors, each one of those was stripped of unneeded outputs and eventually, only the `find_satisfier_and_previous_satisfier_level` method was publicly exposed. But I kept the `find_satisfier` and `find_previous_satisfier` names mainly because naming is hard XD.

In order to find a good name, I'll try summarizing what those functions do. The purpose of `find_satisfier` is to find the oldest decision that caused the current conflict we are trying to solve. Let's say we order assignements in the history of our partial solution like this:

```txt
decision 1, derivation 1a, derivation 1b, decision 2, derivation 2a, decision 3, derivation 3a
```

And let's say we have the incompatibility `{p: 1, q: 1}` that was just created and that we just discovered is satisfied by our current partial solution. We now have to backtrack, but the question is how much? And the purpose of `find_satisfier` is to find out which assignments are irrelevant and what are the oldest assignments relevant to this incompatibility. It turns out for example that decision 1 was `p: 1` and derivation 2a was `q: 1`. If all assignments between decision 1 and derivation 2a are irrelevant, then derivation 2a is our satisfier. The function would then return `(DecisionLevel(2), derivation_2a)` for the satisfier. It also returns the slice of all assignements previous to this one so `&[decision_1 .. decision_2]` if you allow the inaccurate notation.

If we were to return those two things in a struct (name to be found), accurate (verbose) names of the fields could be

```rust
struct Tbf {
    satisfier: (DecisionLevel, Assignment<P,V>),
    history_previous_to_satisfier: &[(DecisionLevel, Assignment<P,V>)],
}
```

where the pair above could be replaced by the `AssignmentRecord` variation we discussed in a previous comment.

**Comment by Eh2406** [comment]
What is this PR waiting on, is it me? If so to share my opinion on what questions?

**Comment by mpizenberg** [comment]
I gave my point of view on the choices for the types `AssignmentDepth` and `AssignmentHistory` in my last two comments. Would love to have an answer from alex or you on thos points

**Comment by Eh2406** [comment]
> DecisionLevel(u32)

Big +1, great improvement.

> history: Vec<(DecisionLevel, Assignment<P, V>)>

Seames good to me. If we need to hide the type more I'd rather see a `history: History<P, V>` over a `history: Vec<AssignmentRecord<P, V>>`, as I could imagine other ways to store this... Edit: removed as it is off topic, and if we are adding a `AssignmentRecord` for other reason then use it.

> find_satisfier_and_previous_satisfier_level 

This section of the code seams under polished in general. (To be clear that is not a complaint. make it work make it clean, make it fast. I never got to the the first one.) As such I am good with incremental improvements. If `AssignmentRecord` makes things clearer for now, wonderfull. If we need to turn that lint off for those functions that is also good.
If we are pulling it out into a struct mybee `SatisfierContext`, it is used all at the bottom of the one file so it doesn't have to be perfectly memorable.

**Comment by mpizenberg** [comment]
Thanks for the feedback! @aleksator when you have time for that, pick up whatever you like from my last comments (or not) and you can merge this. As Jacob says, it's not the most polished part of the code and will eventually need some refinement later anyway.

**Comment by Eh2406** [comment]
> ```rust
> struct Tbf {
>     satisfier: (DecisionLevel, Assignment<P,V>),
>     history_previous_to_satisfier: &[(DecisionLevel, Assignment<P,V>)],
> }
> ```
If I am reading the code wright `satisfier` is always `all_assignments[history_previous_to_satisfier.len()]` and `history_previous_to_satisfier` is always `all_assignments[0..history_previous_to_satisfier.len()]`, if that is true then it can be simplified by returning the slice with `satisfier` being the last element, or even just returning the index of `satisfier`.
Edit: not to say that that should or has to happen in this PR, just a thing I noticed.

**Comment by mpizenberg** [comment]
> If I am reading the code wright satisfier is always all_assignments[history_previous_to_satisfier.len()] and history_previous_to_satisfier is always all_assignments[0..history_previous_to_satisfier.len()]

That's right. We could just return the index, or return the satisfier and the index. Haven't tried those options so I don't know if the code would be clearer or not that way.

**Comment by mpizenberg** [comment]
Renamed `AssignmentDepth` to `DatedAssignment` since "dated" fits well with "history".

Changed `AssignmentHistory` struct to the more verbose but more adequate:

```rust
pub struct SatisfierAndPreviousHistory<'a, P: Package, V: Version> {
    satisfier: DatedAssignment<P, V>,
    previous_history: &'a [DatedAssignment<P, V>],
}
```

**Comment by mpizenberg** [comment]
All changes here were already validated except the two small name changes I just did. They fit better than the previous names but if they are still not perfect, we can change those later. For the time being I'll just go ahead and merge this.

**Pull request merged**
